kernel/sched: Fix edge case in MetaIRQ preemption of cooperative threads

andyross opened the pull request:
When a MetaIRQ preempts a cooperative thread, that thread would be
added back to the generic run queue.  When the MetaIRQ is done, the
highest priority thread will be selected to run, which may obviously
be a cooperative thread of a higher priority than the one that was
preempted.

But that's wrong, because the original thread was promised that it
would NOT be preempted until it reached a scheduling point on its own
(that's the whole point of a cooperative thread, of course).

We need to track the thread that got preempted (one per CPU) and
return to it instead of whatever else the scheduler might have found.

Fixes #20255

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix edge case in MetaIRQ preemption of cooperative threads

When a MetaIRQ preempts a cooperative thread, that thread would be
added back to the generic run queue.  When the MetaIRQ is done, the
highest priority thread will be selected to run, which may obviously
be a cooperative thread of a higher priority than the one that was
preempted.

But that's wrong, because the original thread was promised that it
would NOT be preempted until it reached a scheduling point on its own
(that's the whole point of a cooperative thread, of course).

We need to track the thread that got preempted (one per CPU) and
return to it instead of whatever else the scheduler might have found.

Fixes #20255

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

tyrantlizard77 reviewed (commented):
LGTM.
I'd really like a test case that proves that this works though, and covers the new functions added.

andyross commented:
@wopu-ot already has a test written as part of #20662 actually.  For sure that needs to be split out and added.

wopu-ot commented:
Could a preempted cooperative thread migrate to a different CPU? If that is the case, the metairq_preempted field would not be reset and `next_up` would see stale information.

andyross commented:
> Could a preempted cooperative thread migrate to a different CPU? 

It cannot.  The preempted thread never gets requeued, so it will never be seen by the scheduler logic until it gets picked up on metairq exit.

wopu-ot reviewed (changes requested):
LGTM, I will rework #20662 to only contain the test case

andyross commented:
Realized this morning there was a missed edge case: if a thread gets suspended while a metairq is running and the metairq returns to a different thread, the existing metairq_preempted field would be left non-NULL and that thread, if enabled later, would get incorrectly run on exit from a future metairq even though it hadn't been preempted.

Fixed, and cleaned up the nullification of that field so that it always happens on metairq exit (or if the existing preempted thread is detected to be non-runnable) regardless of what thread might have been stored.

wopu-ot commented:
Maybe I am missing something, but the updated commit looks just like to original commit to me?

andyross commented:
Oops, must have missed a git add.  Should be fixed.  For review, this is the diff from the original: 

```patch
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -182,6 +182,8 @@ static ALWAYS_INLINE struct k_thread *next_up(void)
        if (mirqp != NULL && (th == NULL || !is_metairq(th))) {
                if (!z_is_thread_prevented_from_running(mirqp)) {
                        th = mirqp;
+               } else {
+                       _current_cpu->metairq_preempted = NULL;
                }
        }
 #endif
@@ -319,7 +321,7 @@ static void update_metairq_preempt(struct k_thread *th)
        if (is_metairq(th) && !is_metairq(_current) && !is_preempt(_current)) {
                /* Record new preemption */
                _current_cpu->metairq_preempted = _current;
-       } else if (th == _current_cpu->metairq_preempted) {
+       } else if (!is_metairq(th)) {
                /* Returning from existing preemption */
                _current_cpu->metairq_preempted = NULL;
        }
```

wopu-ot reviewed (approved):
Looking even better after the update.
